File: 3434-find-the-number-of-distinct-colors-among-the-balls/3434-find-the-number-of-distinct-colors-among-the-balls.py
class Solution:
    def queryResults(self, limit: int, queries: List[List[int]]) -> List[int]:
        distinct_colors = {}
        balls = {}
        ans = []

        for num, color in queries:
            if num in balls:
                prev_color = balls[num]
                distinct_colors[prev_color] -= 1
                if distinct_colors[prev_color] == 0:
                    del distinct_colors[prev_color]

            balls[num] = color
            distinct_colors[color] = distinct_colors.get(color, 0) + 1
            ans.append(len(distinct_colors))

        return ans

File: 3434-find-the-number-of-distinct-colors-among-the-balls/3434-find-the-number-of-distinct-colors-among-the-balls.cpp
class Solution {
public:
    vector<int> queryResults(int limit, vector<vector<int>>& queries) {
        unordered_map<int, int> distinct_colors;
        unordered_map<int, int> balls;
        vector<int> ans;
        int n = queries.size();

        for(auto query : queries)
        {
            int num = query[0];
            int color = query[1];

            //balls colored
            if(balls.find(num) != balls.end())
            {
                int prevColor = balls[num];
                distinct_colors[prevColor]--;
                if(distinct_colors[prevColor] == 0) distinct_colors.erase(prevColor);
            }
            balls[num] = color;
            distinct_colors[color]++;
            ans.push_back(distinct_colors.size());
        }
        return ans;
    }
};